README.md
This repository contains the code and supporting files to run TurtleBot 2 demos using ROS 2. Due to reliance on existing Linux-only code and dependencies, these demos are intended for use only on Linux (that could change in the future).

The followwing instructions ar for ROS Bouncy, if you are using ROS Ardent please refer to [these instructions](https://github.com/ros2/turtlebot2_demo/blob/ardent/README.md).

<!-- Here's a video of the very first successful run of ROS 2 follower: https://www.youtube.com/watch?v=YTlls9yHZog.

This is very much a work in progress and there's plenty of functionality present in the ROS 1 equivalent system that's currently missing or disabled, but it's an important step. -->

This demo assumes that you have an Orbbec Astra depth camera. Extra work would be required to use the Kinect or Asus Xtion Pro. Without an Astra, you can still do joystick teleop.
For instructions on how to setup your turtlebot please see [Setup your turtlebot2](https://github.com/ros2/turtlebot2_demo/blob/master/Turtlebot_Setup.md)

# Installation

## Installation from binaries

First, install ROS2 from binaries following [these instructions](https://github.com/ros2/ros2/wiki/Linux-Install-Debians)

Then install the turtlebo2 demo specific packages:
```
export ROS1_DISTRO=melodic
sudo apt install ros-bouncy-turtlebot2* ros-$ROS1_DISTRO-kobuki-ftdi
```

## Installation from source
This assumes that you have ROS Kinetic installed (or at lease have the ros apt repository in your sources)

First, install ROS2 from source following [these instructions](https://github.com/ros2/ros2/wiki/Linux-Development-Setup)

Then get the turtlebo2 demos specific code:
```
cd <YOUR_ROS2_WORKSPACE>
wget https://raw.githubusercontent.com/ros2/turtlebot2_demo/bouncy/turtlebot2_demo.repos
vcs import src < turtlebot2_demo.repos
```

### Install some dependencies:
```bash
export ROS1_DISTRO=melodic # or kinetic if using Ubuntu Xenial
sudo apt-get install --no-install-recommends -y libboost-iostreams-dev libboost-regex-dev libboost-system-dev libboost-thread-dev libceres-dev libgoogle-glog-dev liblua5.2-dev libpcl-dev libprotobuf-dev libsdl1.2-dev libsdl-image1.2-dev libsuitesparse-dev libudev-dev libusb-1.0-0-dev libyaml-cpp-dev protobuf-compiler python-sphinx ros-$ROS1_DISTRO-catkin ros-$ROS1_DISTRO-kobuki-driver ros-$ROS1_DISTRO-kobuki-ftdi
```

Reason for each dependency:
* `ros-$ROS1_DISTRO-kobuki-driver` : our ROS 2 kobuki driver builds on top of this package (and its dependencies)
* `ros-$ROS1_DISTRO-kobuki-ftdi` : we use a `udev` rule from this package
* `ros-$ROS1_DISTRO-common-msgs` : to support use of the `ros1_bridge`, we need the ROS 1 messages available (TODO: document use of the bridge to view depth images and other stuff)

### Build the ros2 code

For resource constrained platforms we will split the build into 2 steps to make sure not to overflow the memory
```bash
colcon build --symlink-install --packages-skip cartographer cartographer_ros cv_bridge opencv_tests ros1_bridge turtlebot2_amcl turtlebot2_drivers turtlebot2_follower turtlebot2_cartographer turtlebot2_teleop vision_opencv
```

Now the resource intensive packages and the ones depending on ROS1 packages:
```bash
source /opt/ros/$ROS1_DISTRO/setup.bash
colcon build --symlink-install --packages-select cartographer cartographer_ros turtlebot2_amcl turtlebot2_cartographer turtlebot2_drivers turtlebot2_follower turtlebot2_teleop
```
Go grab a coffee (or a meal if you compile on ARM)

# Configure a couple of things

## Setup the udev rules

### Copy the astra udev rules

#### If you installed from binaries
If you installed from binary you'll need to download the udev rule by hand:
```bash
wget https://raw.githubusercontent.com/ros2/ros_astra_camera/ros2/56-orbbec-usb.rules
```

#### If you installed from source
```bash
cd ~/ros2_ws/src/ros2/ros_astra_camera
```

#### Copy the rules file
```bash
sudo cp 56-orbbec-usb.rules /etc/udev/rules.d
```

### Copy the kobuki udev rule

```bash
sudo cp `rospack find kobuki_ftdi`/57-kobuki.rules /etc/udev/rules.d
```

### Restart the udev service
```bash
sudo service udev reload
sudo service udev restart
```

## Source your workspace

If installed from Debian packages
```bash
source /opt/ros/$ROS1_DISTRO/setup.bash
source /opt/ros/bouncy/setup.bash
```

If installed from source
```bash
source /opt/ros/$ROS1_DISTRO/setup.bash
source ~/ros2_ws/install/local_setup.bash
```

You'll need to do this step for every terminal you use for these demos

# Run the demos

## Joystick teleop
This is a classic teleoperation demo where the robot can be driven around using a gamepad controller. Thie demo has been tested with logitech controllers and uses `RB` as a deadman, the left joystick for driving forward/backward and the right joystick for rotation.
Try the launch file:
```
launch `ros2 pkg prefix turtlebot2_teleop`/share/turtlebot2_teleop/launch/turtlebot_joy.py
```

Or, run the nodes separately:
```
ros2 run turtlebot2_drivers kobuki_node
```
```
ros2 run joy joy_node
```

Note: this demo assumes that your controller is in D mode (switch on the back) and that the MODE led is on. 

## Follower
This demo uses the astra camera to detect blobs in the depthimage and follow them
Try the launch file:
```
launch `ros2 pkg prefix turtlebot2_follower`/share/turtlebot2_follower/launch/turtlebot_follow.py
```

Or, run the nodes separately.
```
ros2 run turtlebot2_drivers kobuki_node
```
```
ros2 run astra_camera astra_camera_node -- -dw 320 -dh 240 -C -I
```
```
ros2 run turtlebot2_follower follower
```

## Cartographer (mapping)
This demo is using Google cartographer to build a map of the environment. The resulting map can be visualized in RViz using the ros1_bridge (more information below).

### Run the demo
Try the launch file:
```
launch `ros2 pkg prefix turtlebot2_cartographer`/share/turtlebot2_cartographer/launch/turtlebot_carto_2d.py
```

### Visualize the results
The created map can be visualized in Rviz on a remote computer by using the dynamic bridge that converts messages between ROS1 and ROS2.
This assumes that you have a ROS2 dynamic bridge on your system.

#### Installing the bridge

##### From binaries:
Setup your sources as explained on the [setup sources section](https://github.com/ros2/ros2/wiki/Linux-Install-Debians#setup-sources) and then run
```bash
sudo apt update && sudo apt install ros-bouncy-ros1-bridge
```

##### From source:
Build your ROS2 workspace as explained in [these instructions](https://github.com/ros2/ros1_bridge/blob/master/README.md#build-the-bridge-from-source).

#### Run the bridge

Terminal A:
```bash
. /opt/ros/$ROS1_DISTRO/setup.bash
roscore
```

Terminal B:
```bash
. /opt/ros/$ROS1_DISTRO/setup.bash
. <YOUR_ROS2_WORKSPACE>
ros2 run ros1_bridge dynamic_bridge
```

Terminal C:
```bash
. /opt/ros/$ROS1_DISTRO/setup.bash
rosrun rviz rviz
```
Topics you can visualize in Rviz:
- the map on the topic `/map`
- the transforms on the topic `/tf`
- the depth images on the topic `/depth`
- the laserscans on the topic `/scans`

## AMCL (localization)
See the [AMCL demo README](https://github.com/ros2/turtlebot2_demo/blob/master/turtlebot2_amcl/README.md)

# Discussion
What's happening here compared to the ROS 1 versions of these demos? Well, it's 100% ROS 2, with no bridge or shim. We took 4 different
approaches in building the different pieces:

1. Kobuki driver: we wrote a new, very small rclcpp node that calls into the existing kobuki driver packages, which are organized to be roscpp-independent. In this case, we're building on top of ROS 1 packages, but they don't use `roscpp` or other parts of the ROS 1 middleware, so we're just using them as supporting libraries.
2. Astra driver: we forked and ported the existing ROS 1 package (there's no roscpp-independent package separation).
3. Joystick driver: we wrote a simple rclcpp node from scratch (Linux-only for now).
4. Follower node: we created a new package into which we copied and then ported the ROS 1 follower nodelet.

As we start migrating more code to ROS 2, we'll discover more about these kinds of techniques and arrive at some best practices that we can recommend for similar projects.


turtlebot2_amcl/CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package turtlebot2_amcl
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

0.4.1 (2018-07-20)
------------------

0.4.0 (2017-12-08)
------------------
* add missing maintainer names (`#73 <https://github.com/ros2/turtlebot2_demo/issues/73>`_)
* 0.0.3
* bix broken links pointing to amcl_readme branch (`#58 <https://github.com/ros2/turtlebot2_demo/issues/58>`_)
* Update for amcl being installed to libexec now (`#54 <https://github.com/ros2/turtlebot2_demo/issues/54>`_)
* update README (`#21 <https://github.com/ros2/turtlebot2_demo/issues/21>`_)
* 0.0.2
* install parameter bridge config too (`#49 <https://github.com/ros2/turtlebot2_demo/issues/49>`_)
* Add amcl demo (`#44 <https://github.com/ros2/turtlebot2_demo/issues/44>`_)
* Contributors: Dirk Thomas, Mikael Arguedas, dhood


turtlebot2_amcl/launch/turtlebot_amcl.py
# Copyright 2017 Open Source Robotics Foundation, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import argparse
import os
import sys

from ament_index_python.packages import get_package_share_directory
from launch import LaunchDescriptor
from launch.exit_handler import restart_exit_handler
from launch.launcher import DefaultLauncher
from launch.output_handler import ConsoleOutput
from ros2run.api import get_executable_path


def launch(launch_descriptor, parser, argv):
    #parser = argparse.ArgumentParser(description='launch amcl turtlebot demo')
    parser.add_argument(
        '--map',
        help='path to map (will be passed to map_server)')
    args = parser.parse_args(argv)

    ld = launch_descriptor

    package = 'turtlebot2_drivers'
    ld.add_process(
        cmd=[get_executable_path(package_name=package, executable_name='kobuki_node')],
        name='kobuki_node',
        exit_handler=restart_exit_handler,
    )
    package = 'astra_camera'
    ld.add_process(
        cmd=[
            get_executable_path(package_name=package, executable_name='astra_camera_node'),
            '-dw', '320', '-dh', '240', '-C', '-I'],
        name='astra_camera_node',
        exit_handler=restart_exit_handler,
    )
    package = 'depthimage_to_laserscan'
    ld.add_process(
        cmd=[
            get_executable_path(
                package_name=package, executable_name='depthimage_to_laserscan_node')],
        name='depthimage_to_laserscan_node',
        exit_handler=restart_exit_handler,
    )
    package = 'tf2_ros'
    ld.add_process(
        # The XYZ/Quat numbers for base_link -> camera_rgb_frame are taken from the
        # turtlebot URDF in
        # https://github.com/turtlebot/turtlebot/blob/931d045/turtlebot_description/urdf/sensors/astra.urdf.xacro
        cmd=[
            get_executable_path(
                package_name=package, executable_name='static_transform_publisher'),
            '-0.087', '-0.0125', '0.287',
            '0', '0', '0', '1',
            'base_link',
            'camera_rgb_frame'
        ],
        name='static_tf_pub_base_rgb',
        exit_handler=restart_exit_handler,
    )
    package = 'tf2_ros'
    ld.add_process(
        # The XYZ/Quat numbers for camera_rgb_frame -> camera_depth_frame are taken from the
        # turtlebot URDF in
        # https://github.com/turtlebot/turtlebot/blob/931d045/turtlebot_description/urdf/sensors/astra.urdf.xacro
        cmd=[
            get_executable_path(
                package_name=package, executable_name='static_transform_publisher'),
            '0', '0.0250', '0',
            '0', '0', '0', '1',
            'camera_rgb_frame',
            'camera_depth_frame'
        ],
        name='static_tf_pub_rgb_depth',
        exit_handler=restart_exit_handler,
    )
    package = 'joy'
    ld.add_process(
        cmd=[get_executable_path(package_name=package, executable_name='joy_node')],
        name='joy_node',
        exit_handler=restart_exit_handler,
    )
    package = 'teleop_twist_joy'
    ld.add_process(
        cmd=[get_executable_path(package_name=package, executable_name='teleop_node')],
        name='teleop_node',
        exit_handler=restart_exit_handler,
    )
    turtlebot2_amcl_share = get_package_share_directory('turtlebot2_amcl')
    map_path = os.path.join(turtlebot2_amcl_share, 'examples', 'osrf_map.yaml')
    if args.map:
        map_path = args.map
    package = 'map_server'
    ld.add_process(
        cmd=[get_executable_path(package_name=package, executable_name='map_server'), map_path],
        name='map_server',
    )
    ld.add_process(
        cmd=[
            'amcl', '--use-map-topic',
        ],
        name='amcl',
        exit_handler=restart_exit_handler,
        output_handlers=[ConsoleOutput()],
    )

    return ld


def main(argv=sys.argv[1:]):
    launcher = DefaultLauncher()
    launch_descriptor = launch(LaunchDescriptor(), argv)
    launcher.add_launch_descriptor(launch_descriptor)
    rc = launcher.launch()
    return rc


if __name__ == '__main__':
    sys.exit(main())
